File: app/src/main/java/tdc/com/bidirectionaldijkstrasalgorithm/model/DijkstraExplorer.java
package tdc.com.bidirectionaldijkstrasalgorithm.model;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

class DijkstraExplorer {
  // visited, means we visited this node during exploring a node, we haven't fully explored this
  // node yet.
  Map<String, String> visitedMap;

  // explored, means we have fully explored this node, we have been to all its neighbors nodes.
  Set<String> exploredSet;

  // the shortest distance of a node to the init node.
  Map<String, Double> distanceMap;
  PriorityQueue<Edge> pq;
  Graph graph;

  public DijkstraExplorer(String initNode, Graph graph) {
    int n = graph.size();
    this.graph = graph;
    this.exploredSet = new HashSet<>(n);
    this.visitedMap = new HashMap<>(n);
    this.distanceMap = new HashMap<>(n);
    this.pq = new PriorityQueue<>((a, b) -> Double.compare(a.distance, b.distance));
    this.distanceMap.put(initNode, 0.0);
    visitedMap.put(initNode, null);
    this.pq.offer(new Edge(initNode, 0.0));
  }

  public String exploring(DijkstraExplorer waitingExplorer, Set<String> commonSet) {
    String exploringNode = null;
    if (pq.size() > 0) {
      Edge poll = pq.poll();
      exploringNode = poll.to;
      if (exploredSet.add(exploringNode)) {
        List<Edge> edgeList = graph.getEdges(exploringNode);
        if (edgeList != null) {
          for (Edge edge : edgeList) {
            String neighbor = edge.to;
            if (!exploredSet.contains(neighbor)) {
              double newCostToNeighbor = edge.distance + distanceMap.get(exploringNode);
              if (newCostToNeighbor < distanceMap.getOrDefault(neighbor, Double.MAX_VALUE)) {
                distanceMap.put(neighbor, newCostToNeighbor);
                visitedMap.put(neighbor, exploringNode);
                pq.add(new Edge(neighbor, newCostToNeighbor));
                if (waitingExplorer.distanceMap.containsKey(neighbor)) {
                  commonSet.add(neighbor);
                }
              }
            }
          }
        }
      }
    }
    return exploringNode;
  }
}

File: app/src/main/java/tdc/com/bidirectionaldijkstrasalgorithm/model/forkotlin/DijkstraExplorer.kt
package tdc.com.bidirectionaldijkstrasalgorithm.model.forkotlin

import java.util.*

class DijkstraExplorer(initNode: String, graph: Graph) {
    // visited, means we visited this node during exploring a node, we haven't fully explored this
    // node yet.
    val visitedMap: MutableMap<String, String?>

    // explored, means we have fully explored this node, we have been to all its neighbors nodes.
    val exploredSet: MutableSet<String>

    // the shortest distance of a node to the init node.
    val distanceMap: MutableMap<String, Double>
    private val pq: PriorityQueue<Edge>
    private val graph: Graph
    fun exploring(waitingExplorer: DijkstraExplorer, commonSet: MutableSet<String>): String? {
        var exploringNode: String? = null
        if (pq.size > 0) {
            val poll = pq.poll()
            exploringNode = poll!!.to
            if (exploredSet.add(exploringNode)) {
                val edgeList = graph.getEdges(exploringNode)
                if (null != edgeList) {
                    for (edge in edgeList) {
                        val neighbor = edge.to
                        if (neighbor !in exploredSet) {
                            val newCostToNeighbor = edge.distance + distanceMap[exploringNode]!!
                            if (newCostToNeighbor < distanceMap.getOrDefault(
                                    neighbor,
                                    Double.MAX_VALUE
                                )
                            ) {
                                distanceMap[neighbor] = newCostToNeighbor
                                visitedMap[neighbor] = exploringNode
                                pq.add(Edge(neighbor, newCostToNeighbor))
                                if (waitingExplorer.distanceMap.containsKey(neighbor)) {
                                    commonSet.add(neighbor)
                                }
                            }
                        }
                    }

                }
            }
        }
        return exploringNode
    }

    init {
        val n = graph.size()
        this.graph = graph
        exploredSet = HashSet(n)
        visitedMap = HashMap(n)
        distanceMap = HashMap(n)
        pq =
            PriorityQueue { a, b ->
                a.distance.compareTo(b.distance)
            }
        distanceMap[initNode] = 0.0
        visitedMap[initNode] = null
        pq.offer(Edge(initNode, 0.0))
    }
}